### PR TITLE
Improve tasks CLI

### PR DESCRIPTION
**Pull request by etpinard**
resolves https://github.com/plotly/plotly.js/issues/794

### In brief

this PR:

- rewrote all the remaining shell script found in `tasks/` in node which (1) will allow to Windows users to run them hassle-free (2) allow us to DRY up significant amounts of tasks code.
- refactored the `pretest` lifecycle task. It now contains everything needed to prep up the CircleCI box before testing including booting up the imagetest docker container (which cleans up our `circle.yml` significantly)
- hard-coded the mapbox access token used for testing into `tasks/util/constants.js`. No more `export <mapbox-access-token>` needed to start dev'ing!
- the other commits DRYed up the existing `tasks/` and `devtools/` code.

### Ideas for future development

- Drop `docker-compose` for booting the imagetest container locally. That would mean one less dependency for devs :tada: . Use `npm run pretest` instead (and maybe add an `npm run postest` task to shut down the container while at it).
- Factor out all bundling code into a separate module (see https://github.com/plotly/plotly.js/issues/768)

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -9,20 +9,17 @@ machine:
...
-    - docker run -d --name mytestbed -v $PWD:/var/www/streambed/image_server/plotly.js -p 9010:9010 plotly/testbed:latest
-    - sudo ./tasks/run_in_testbed.sh mytestbed "cp -f test/image/index.html ../server_app/index.html"
-    - wget --server-response --spider --tries=8 --retry-connrefused http://localhost:9010/ping
+
 test:
   override:
-    - sudo ./tasks/run_in_testbed.sh mytestbed "export CIRCLECI=1 && node test/image/compare_pixels_test.js"
-    - sudo ./tasks/run_in_testbed.sh mytestbed "node test/image/export_test.js"
+    - npm run test-image
+    - npm run test-export
```
Comment (etpinard) [inline circle.yml]: the local vs CI differences are now handled at the task-runner script level.

File: circle.yml [hunk 2]
```diff
@@ -9,20 +9,17 @@ machine:
...
-
 dependencies:
   pre:
     - docker pull plotly/testbed:latest
   post:
     - npm run cibuild
     - npm run pretest
-    - docker run -d --name mytestbed -v $PWD:/var/www/streambed/image_server/plotly.js -p 9010:9010 plotly/testbed:latest
-    - sudo ./tasks/run_in_testbed.sh mytestbed "cp -f test/image/index.html ../server_app/index.html"
-    - wget --server-response --spider --tries=8 --retry-connrefused http://localhost:9010/ping
```
Comment (etpinard) [inline circle.yml]: all these :arrow_double_up: are now part of `npm run pretest`

File: package.json [hunk 3]
```diff
@@ -27,21 +27,21 @@
...
+    "test-image": "node tasks/test_image.js",
+    "test-export": "node tasks/test_export.js",
+    "test-syntax": "node tasks/test_syntax.js",
     "test-bundle": "node tasks/test_bundle.js",
     "test": "npm run citest-jasmine && npm run test-image && npm run test-syntax && npm run test-bundle",
     "start-test_dashboard": "node devtools/test_dashboard/server.js",
     "start-image_viewer": "node devtools/image_viewer/server.js",
     "start": "npm run start-test_dashboard",
-    "baseline": "./tasks/baseline.sh",
+    "baseline": "node tasks/baseline.js",
```
Comment (mdtusz) [inline package.json]: _Who needs `make` when you have npm!_

File: tasks/util/common.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+var fs = require('fs');
+var exec = require('child_process').exec;
+
```
Comment (mdtusz) [inline tasks/util/common.js]: 🎉  great module!

File: tasks/preprocess.js [hunk 5]
```diff
@@ -1,35 +1,45 @@
...
+        pullFontSVG(data.toString(), constants.pathToFontSVGBuild);
+    });
+}
 
-// inject package version into source index files
-updateVersion(constants.pathToPlotlyCore);
-updateVersion(constants.pathToPlotlyGeoAssetsSrc);
+// copy topojson files from sane-topojson to dist/
+function copyTopojsonFiles() {
+    fs.copy(constants.pathToTopojsonSrc, constants.pathToTopojsonDist,
```
Comment (mdtusz) [inline tasks/preprocess.js]: 🐄  separate lines for each arg?

File: tasks/util/constants.js [hunk 6]
```diff
@@ -67,6 +67,7 @@ module.exports = {
     pathToJasmineTests: path.join(pathToRoot, 'test/jasmine/tests'),
     pathToJasmineBundleTests: path.join(pathToRoot, 'test/jasmine/bundle_tests'),
 
+    mapboxAccessToken: 'pk.eyJ1IjoiZXRwaW5hcmQiLCJhIjoiY2luMHIzdHE0MGFxNXVubTRxczZ2YmUxaCJ9.hwWZful0U2CQxit4ItNsiQ',
```
Comment (mdtusz) [inline tasks/util/constants.js]: Maybe we should add a comment that _we're aware there's a token here and that we're ok with it_ so we don't have too many overzealous contributions to improve security?

File: tasks/util/container_commands.js [hunk 7]
```diff
@@ -52,10 +52,11 @@ function getRunCI(commands) {
     commands = ['export CIRCLECI=1', containerCommands.cdHome].concat(commands);
 
     return [
+        'sudo',
         'lxc-attach -n',
-        '$(docker inspect --format \'{{.Id}}\'' + constants.testContainerName + ')',
+        '$(docker inspect --format \'{{.Id}}\' ' + constants.testContainerName + ')',
         '-- bash -c',
-        commands.join(' && ')
+        '"' + commands.join(' && ') + '"'
```
Comment (mdtusz) [inline tasks/util/container_commands.js]: 🐄  may as well move the `.join` to where commands is redefined above.

**Comment by mdtusz** [comment]
Great PR! 

After those little cows and things, 💃 !

File: tasks/util/container_commands.js [hunk 8]
```diff
@@ -15,6 +15,7 @@ containerCommands.ping = [
 containerCommands.setup = [
     containerCommands.cpIndex,
     containerCommands.restart,
+    'sleep 1',
```
Comment (etpinard) [inline tasks/util/container_commands.js]: Looks like the `npm run pretest` is a little more flaky than the previous list of shell commands on CIrcleCi. 

Not sure what's up.

I've added sleep delay:
- between the `nw` restart and the ping as well as as
- between the docker run and setup commands. 

Hopefully that will fix it. But we might have to look into this further.

**Pull request merged**